src/passes/OptimizeInstructions.cpp

```diff
@@ -1345,15 +1377,57 @@ struct OptimizeInstructions
         }
       }
       // operations on all 1s
-      // TODO: shortcut method to create an all-ones?
       if (right->value == Literal(int32_t(-1)) ||
           right->value == Literal(int64_t(-1))) {
         if (binary->op == Abstract::getBinary(type, Abstract::And)) {
+          // x & -1   ==>   x
           return binary->left;
         } else if (binary->op == Abstract::getBinary(type, Abstract::Or) &&
                    !EffectAnalyzer(getPassOptions(), features, binary->left)
                       .hasSideEffects()) {
+          // x & -1   ==>   -1
           return binary->right;
+        } else if (binary->op == Abstract::getBinary(type, Abstract::Sub)) {
+          // x - (-1)   ==>   x + 1
+          Builder builder(*getModule());
+          return builder.makeBinary(
+            Abstract::getBinary(type, Abstract::Add),
+            binary->left,
+            builder.makeConst(Literal::makeFromInt32(1, type)));
+        } else if ((binary->op == Abstract::getBinary(type, Abstract::RemS) ||
+                    binary->op == GtUInt32 || binary->op == GtUInt64) &&
```

Comment:
for consistency we can use `Abstract` here too (looks like we need to add it to `src/ir/abstract.h`)

Reply:
(alternatively, using more templating might be better)